### PR TITLE
Hide password in -stdin `walletpassphrase`

### PR DESCRIPTION
**Pull request by oxarbitrage**
An approach for issue https://github.com/zcash/zcash/issues/4932

Code in this PR will hide the first line of input when we are in `-stdin` mode and only when the `walletpassphrase` command is used. This works for me but i understand it can be confusing, open to other options.

Without this PR:

```
$ ./src/zcash-cli -stdin walletpassphrase\n
mypass\n
120\n
<Ctrl+D>
$
```

With PR:

```
$ ./src/zcash-cli -stdin walletpassphrase\n
[here we type the password and enter but this is not shown]
120\n
<Ctrl+D>
$
```

Function `SetStdinEcho` is not mine(https://stackoverflow.com/a/1455007). It works in linux, it should work in windows, not sure about mac.

This PR also moves the `-stdin` option to the Options section of the cli help.

------------
Please ensure this checklist is followed for any pull requests for this repo. This checklist must be checked by both the PR creator and by anyone who reviews the PR.
* [ ] Relevant documentation for this PR has to be completed and reviewed by @mdr0id before the PR can be merged
* [ ] A test plan for the PR must be documented in the PR notes and included in the test plan for the next regular release

As a note, all buildbot tests need to be passing and all appropriate code reviews need to be done before this PR can be merged

### CODE REVIEW

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -275,10 +308,21 @@ int CommandLineRPC(int argc, char *argv[])
         }
         std::vector<std::string> args = std::vector<std::string>(&argv[1], &argv[argc]);
         if (GetBoolArg("-stdin", false)) {
+            bool hide = false;
+            if (strncmp(argv[1], "walletpassphrase", 16) == 0) {
```
Comment (daira) [inline src/bitcoin-cli.cpp:312]:
```suggestion
            if (argc > 1 && args[0] == "walletpassphrase") {
```

**Review by daira — CHANGES_REQUESTED** [review]
Change needed.

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
     return reply;
 }
 
+void SetStdinEcho(bool enable = true)
+{
+#ifdef WIN32
+    HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
```
Comment (daira) [inline src/bitcoin-cli.cpp:276]:
```suggestion
    HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
    assert(hStdin != INVALID_HANDLE_VALUE && hStdin != NULL);
```

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
...
+    DWORD mode;
+    GetConsoleMode(hStdin, &mode);
```
Comment (daira) [inline src/bitcoin-cli.cpp:278]:
```suggestion
    assert(GetConsoleMode(hStdin, &mode) != 0);
```

File: src/bitcoin-cli.cpp [hunk 4]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
     return reply;
 }
 
+void SetStdinEcho(bool enable = true)
```
Comment (daira) [inline src/bitcoin-cli.cpp:273]:
```suggestion
bool SetStdinEcho(bool enable = true)
```

File: src/bitcoin-cli.cpp [hunk 5]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
...
+    HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE);
+    DWORD mode;
+    GetConsoleMode(hStdin, &mode);
+
+    if (!enable)
+        mode &= ~ENABLE_ECHO_INPUT;
+    else
+        mode |= ENABLE_ECHO_INPUT;
+
+    SetConsoleMode(hStdin, mode);
```
Comment (daira) [inline src/bitcoin-cli.cpp:285]:
```suggestion
    if (SetConsoleMode(hStdin, mode) == 0) return false;
```

File: src/bitcoin-cli.cpp [hunk 6]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
...
+
+#else
+    struct termios tty;
+    tcgetattr(STDIN_FILENO, &tty);
```
Comment (daira) [inline src/bitcoin-cli.cpp:289]:
```suggestion
    if (tcgetattr(STDIN_FILENO, &tty) != 0) return false;
```

File: src/bitcoin-cli.cpp [hunk 7]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
...
+    if (!enable)
+        tty.c_lflag &= ~ECHO;
+    else
+        tty.c_lflag |= ECHO;
+
+    (void) tcsetattr(STDIN_FILENO, TCSANOW, &tty);
```
Comment (daira) [inline src/bitcoin-cli.cpp:295]:
```suggestion
    if (tcsetattr(STDIN_FILENO, TCSANOW, &tty) != 0) return false;
```

File: src/bitcoin-cli.cpp [hunk 8]
```diff
@@ -263,6 +270,32 @@ UniValue CallRPC(const std::string& strMethod, const UniValue& params)
...
+#endif
```
Comment (daira) [inline src/bitcoin-cli.cpp:296]:
```suggestion
#endif
    return true;
```

File: src/bitcoin-cli.cpp [hunk 9]
```diff
@@ -275,10 +308,21 @@ int CommandLineRPC(int argc, char *argv[])
...
+                hide = true;
+            }
+
             // Read one arg per line from stdin and append
             std::string line;
-            while (std::getline(std::cin,line))
+            while (std::getline(std::cin,line)) {
                 args.push_back(line);
+                if (hide) {
+                    SetStdinEcho(true);
```
Comment (daira) [inline src/bitcoin-cli.cpp:322]:
```suggestion
                    if (!SetStdinEcho(true))
                        throw std::runtime_error("unable to set terminal back to echoing");
```

File: src/bitcoin-cli.cpp [hunk 10]
```diff
@@ -275,10 +308,21 @@ int CommandLineRPC(int argc, char *argv[])
         }
         std::vector<std::string> args = std::vector<std::string>(&argv[1], &argv[argc]);
         if (GetBoolArg("-stdin", false)) {
+            bool hide = false;
+            if (strncmp(argv[1], "walletpassphrase", 16) == 0) {
+                SetStdinEcho(false);
```
Comment (daira) [inline src/bitcoin-cli.cpp:313]:
```suggestion
                if (!SetStdinEcho(false))
                    throw std::runtime_error("unable to set terminal to non-echoing");
```

File: src/bitcoin-cli.cpp [hunk 11]
```diff
@@ -19,6 +19,13 @@
 
 #include <univalue.h>
 
+#ifdef WIN32
+#include <windows.h>
+#else
+#include <termios.h>
+#include <unistd.h>
```
Comment (daira) [inline src/bitcoin-cli.cpp:26]:
```suggestion
#include <unistd.h>
#include <cerrno>
```

**Review by daira — APPROVED** [review]
utACK. Needs one more review (especially of the code added by my suggestions), and testing on Windows, Linux, and macOS. The testing should also cover the case where input is redirected.

**Comment by daira** [comment]
Tested on Windows 8.1, in both `cmd.exe` and Powershell:
* `.\zcash-cli --stdin walletpassphrase` does not echo the first line of input, but does echo the rest.
* `type foo.txt |.\zcash-cli --stdin walletpassphrase` does not complain and takes the first line of `foo.txt` as the passphrase.
* other RPC commands are not affected.

**Comment by daira** [comment]
Also tested on Linux.

**Comment by nuttycom** [comment]
Works fine on OSX.

**Review by nuttycom — APPROVED** [review]
ACK

**Comment by nuttycom** [comment]
@zkbot r+

**Pull request merged**